### PR TITLE
add PyTorchModule helper which embeds an ONNX-exported pytorch module…

### PR DESCRIPTION
**Pull request by zdevito**
… into a Caffe2 model builder

### CODE REVIEW

File: onnx_pytorch/pytorch_helper.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            will become inputs of the exported model, in the order they
+            occur in args.  If args is a Variable, this is equivalent
+            to having called it with a 1-ary tuple of that Variable.
+            (Note: passing keyword arguments to the model is not currently
+            supported.  Give us a shout if you need it.)
+        caffe2_inputs (list of str or caffe2.python.core.BlobReference): the
+           caffe2 Blobs that should be inputs to this network. Must be
+           the same length as sample_arguments
+        prefix_name: prefix name to add to each member of the blob, if None then
+           a freesh prefix pytorch_input_N/ is used
```
Comment (jamesr66a) [inline onnx_pytorch/pytorch_helper.py]: is `freesh` one of those new terms the kids are using these days? like `fleek`?

File: onnx_pytorch/pytorch_helper.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    predict_net = Net(predict_net).Clone('anon', _FakeDict(remap_blob_name))
+    helper.net.AppendNet(predict_net)
+
+    init_net = Net(init_net).Clone('anon', _FakeDict(remap_blob_name))
+    helper.param_init_net.AppendNet(init_net)
+
+    results = tuple([BlobReference(remap_blob_name(x.name), helper.net)
+                     for x in onnx_model.graph.output])
+    if len(results) == 1:
```
Comment (jamesr66a) [inline onnx_pytorch/pytorch_helper.py]: I'm partial toward statically typed languages so idk if i'm particularly fond of this return value pattern. Would this change callsites significantly to just always return a tuple?

File: test/test_pytorch_helper.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        class SuperResolutionNet(nn.Module):
+            def __init__(self, upscale_factor, inplace=False):
+                super(SuperResolutionNet, self).__init__()
+
+                self.relu = nn.ReLU(inplace=inplace)
+                self.conv1 = nn.Conv2d(1, 64, (5, 5), (1, 1), (2, 2))
+                self.conv2 = nn.Conv2d(64, 64, (3, 3), (1, 1), (1, 1))
+                self.conv3 = nn.Conv2d(64, 32, (3, 3), (1, 1), (1, 1))
+                self.conv4 = nn.Conv2d(32, upscale_factor ** 2, (3, 3), (1, 1), (1, 1))
```
Comment (jamesr66a) [inline test/test_pytorch_helper.py:25]: now that's what I call deep learning

**Review by jamesr66a — APPROVED** [review]
LGTM. Thanks!

File: onnx_pytorch/pytorch_helper.py (hunk 2 above)
Comment (zdevito) [reply onnx_pytorch/pytorch_helper.py]: Sure, we could always return a tuple if you'd like, but you would always need to write:

`foo, = PyTorchModule(`

for single returns.

Ill do that unless there are other objections (I am guessing not).

Comment (ezyang) [reply onnx_pytorch/pytorch_helper.py]: FWIW, the PyTorch convention is to not wrap in tuple if it's a single return.

File: onnx_pytorch/pytorch_helper.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        model, sample_arguments,  f, export_params=True)
+    onnx_model = onnx.load(io.BytesIO(f.getvalue()))
+    init_net, predict_net = Caffe2Backend.onnx_graph_to_caffe2_net(
+        onnx_model.graph)
+
+    initialized = set([x.name for x in onnx_model.graph.initializer])
+    uninitialized_inputs = {x.name: i for i, x in enumerate(
+        onnx_model.graph.input) if x.name not in initialized}
+
+    if(len(uninitialized_inputs) != len(caffe2_inputs)):
```
Comment (dzhulgakov) [inline onnx_pytorch/pytorch_helper.py:67]: urgh, we should really return MetaNetDef-style of object from Caffe2 that separates inputs and parameters

File: onnx_pytorch/pytorch_helper.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        raise ValueError('Expected {} inputs but found {}'.format(
+            len(uninitialized_inputs), len(caffe2_inputs)))
+
+    def remap_blob_name(name):
+        if name in uninitialized_inputs:
+            idx = uninitialized_inputs[name]
+            return str(caffe2_inputs[idx])
+        return prefix_name + name
+
+    predict_net = Net(predict_net).Clone('anon', _FakeDict(remap_blob_name))
```
Comment (dzhulgakov) [inline onnx_pytorch/pytorch_helper.py:77]: there's also clone_and_bind_net that can get both prefix and blob_remap. It also supports "input_record", but it's advanced C2 stuff on python and we probably shouldn't go there.

we can try it, but not sure whether we need it

File: test/test_pytorch_helper.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+from torch.autograd import Variable
+import torch.onnx
+import torch.nn.init as init
+from caffe2.python.model_helper import ModelHelper
+from onnx_pytorch.pytorch_helper import PyTorchModule
+import unittest
+from caffe2.python.core import workspace
+
+
+class TestCaffe2Backend(unittest.TestCase):
```
Comment (dzhulgakov) [inline test/test_pytorch_helper.py:13]: nit: rename the test

**Review by dzhulgakov — COMMENTED** [review]
cool stuff

File: onnx_pytorch/pytorch_helper.py (hunk 2 above)
Comment (dzhulgakov) [reply onnx_pytorch/pytorch_helper.py]: usually python convention is to return single value instead of tuple. Most libs (includeing pytorch/c2) follow it

**Pull request merged**